### PR TITLE
fix: avoid auth misclassification on skips

### LINKED ISSUE #282: [Bug] Timeout or Parse failure misreported as API_KEY_INVALID in Council Verdict
## Problem

Reviewer jobs that time out (exit 124) or fail to produce valid JSON are being reported as `API Error: API_KEY_INVALID` in the final PR comment, causing developers to waste time investigating secret configuration when the root cause is model performance or execution timeout.

Confirmed in `misty-step/cerberus-cloud` PR #53 ([run #22399678826](https://github.com/misty-step/cerberus-cloud/actions/runs/22399678826)):
- The `Security` reviewer timed out at 600s (`opencode exit=124`)
- The Council Verdict comment showed: `Security | Auth error (invalid key) | Check that the API key secret is set and valid.`
- Findings included `🔵 API Error: API_KEY_INVALID (api_error)`

The root cause: `looks_like_api_error()` in `scripts/parse-review.py` (lines 399–426) pattern-matches against the raw model output. The word `"authentication"` appears in timeout/error output from certain model providers, tripping the `API_KEY_INVALID` branch instead of returning a clean timeout SKIP.

The ordering in `main()` (lines 1005–1030) does check `detect_timeout` before `detect_api_error` for explicit markers, but `looks_like_api_error` is only called on the no-JSON-block path and is broader than `detect_api_error` — it catches `"authentication"` as a substring match with no context guard.

## Context

- **Vision**: Agentic code review platform — multi-perspective review, auto-triage, and observability. Replace CodeRabbit.
- **Related**: #216 (parse-failure policy), #263 (transient gh parse failures in post-verdict-review)
- `detect_timeout` (lines 261–284) correctly identifies `timeout after Ns` markers and produces a clean SKIP with `category: "timeout"`. The bug is that `looks_like_api_error` can intercept output that contains incidental auth-related strings before the timeout marker check runs on that path, or that the timeout marker is absent and only auth-matching heuristics fire.
- The cross-module string contract between `generate_skip_verdict()` and `post-comment.sh`'s skip-banner renderer (lines 88–100) means the wrong `error_type` string flows all the way into the rendered PR comment.

## Acceptance Criteria

- [ ] Given a reviewer that exits with code 124 (timeout) and produces output containing the word "authentication" (e.g., from a provider error page before timeout), when `parse-review.py` processes the output, then the verdict must have `category: "timeout"` and `summary` containing "timeout", never `category: "api_error"` with `title` matching `API_KEY_INVALID`
- [ ] Given a reviewer that produces no JSON block and whose raw output contains only ambiguous auth-adjacent strings (no timeout marker), when `looks_like_api_error` runs, then the error type emitted must be `API_CREDITS_DEPLETED` or `API_ERROR` only when the evidence is unambiguous — the `"authentication"` pattern alone must not classify as `API_KEY_INVALID`
- [ ] Given a `generate_skip_verdict("RATE_LIMIT", ...)` call, when `post-comment.sh` renders the skip banner, then the banner must not display the API key error message (regression guard for cross-module string contract)

## Affected Files

- `scripts/parse-review.py` — `looks_like_api_error()` (lines 399–426): tighten `"authentication"` pattern to require additional context; ensure `detect_timeout` runs unconditionally before `looks_like_api_error` even on partial-output paths
- `scripts/parse-review.py` — `main()` (lines 1005–1030): verify ordering guarantees hold for all code paths, not just the explicit-marker path
- `tests/test_parse_review.py` — add parametrized test covering timeout output that contains incidental auth strings

## Verification

```bash
cd /Users/phaedrus/Development/cerberus-mono/cerberus

# Run the existing parse-review test suite
python3 -m pytest tests/test_parse_review.py -v

# Targeted regression: simulate timeout output that mentions "authentication"
echo 'Review Timeout: timeout after 600s
authentication error from provider' | python3 scripts/parse-review.py
# Expected: verdict=SKIP, category=timeout, NOT api_error

# Run full suite to verify no regressions
python3 -m pytest tests/ -v -x --timeout=60
```

## Boundaries

- Do NOT change the verdict comment markdown template or post-comment.sh skip-banner logic — the fix must happen upstream in parse-review.py before the wrong error_type propagates
- Do NOT alter `detect_api_error()` behaviour for cases where `"API Error:"` is explicitly present in the output — only `looks_like_api_error()` heuristics are in scope
- Do NOT change the timeout detection path for output that already contains a `Review Timeout:` marker

## Approach

Tighten `looks_like_api_error` to require the `"authentication"` substring to co-occur with an HTTP status code or an explicit auth failure keyword, rather than matching it as a standalone signal:

```python
# Current — too broad:
(r"authentication", "API_KEY_INVALID", "Authentication error"),

# Proposed — require corroboration:
(r"(?:401|403|invalid_api_key|incorrect_api_key).*authentication|authentication.*(?:401|403|failed|invalid)", "API_KEY_INVALID", "Authentication error"),
```

Also add a guard in `main()` to call `detect_timeout` before `looks_like_api_error` on _all_ paths (not just the explicit-marker path), so a partial output containing both a timeout message and incidental auth strings always resolves to timeout first.

Add a parametrized integration test:

```python
@pytest.mark.parametrize("raw,expected_category", [
    ("Review Timeout: timeout after 600s\nauthentication error", "timeout"),
    ("Review Timeout: timeout after 300s\nFast-path: attempted", "timeout"),
    ("401 Unauthorized\nauthentication failed", "api_error"),
])
def test_timeout_beats_auth_heuristic(tmp_path, raw, expected_category, monkeypatch):
    monkeypatch.setenv("PERSPECTIVE", "guard")
    result = subprocess.run(
        ["python3", "scripts/parse-review.py"],
        input=raw, capture_output=True, text=True
    )
    data = json.loads(result.stdout)
    assert data["findings"][0]["category"] == expected_category
```

## Overview

```mermaid
flowchart TD
    A[raw model output] --> B{detect_timeout?}
    B -->|yes| C[SKIP verdict\ncategory: timeout]
    B -->|no| D{detect_api_error\nexplicit marker?}
    D -->|yes| E[SKIP verdict\ncategory: api_error\ntype from explicit marker]
    D -->|no| F{JSON block found?}
    F -->|no| G{looks_like_api_error?}
    G -->|yes - auth pattern alone| H["⚠️ BUG: API_KEY_INVALID\n(should be guarded)"]
    G -->|yes - unambiguous| I[SKIP verdict\ncorrect error_type]
    G -->|no| J[parse-failure SKIP]
    F -->|yes| K[validate + emit verdict]

    style H fill:#f88,stroke:#c00
```

---
Created by `/groom`

### PR DESCRIPTION
**Pull request by moomooskycow**
## Why This Matters

Cerberus was telling users to debug API keys when the real failure mode was a timeout or an unstructured provider response. That wastes time, hides the true root cause, and makes SKIP verdicts harder to trust as an operational signal.

This PR fixes issue #282 by tightening `parse-review.py` so incidental `authentication` text no longer becomes `API_KEY_INVALID`, while explicit timeout markers still win cleanly.

Closes #282

## Trade-offs / Risks

- The auth heuristic is narrower now, so it intentionally prefers generic parse-failure over falsely accusing the API key when evidence is ambiguous.
- I did not change the explicit `API Error:` path or downstream comment templates; the fix stays at the parser boundary where the misclassification starts.
- `make validate` still fails in `ruff` on unrelated pre-existing lint violations outside this diff, so the repo-wide gate is not fully green yet despite the full pytest pass.

## Intent Reference

Intent summary:
- Timeout output that happens to mention `authentication` must still emit a timeout SKIP, never `API_KEY_INVALID`.
- Ambiguous auth-adjacent unstructured output must only classify as auth failure when the evidence is unambiguous.
- Non-auth skip types like `RATE_LIMIT` must not point operators at API key debugging.

Source issue: https://github.com/misty-step/cerberus/issues/282

## Changes

- Tightened `looks_like_api_error()` in `scripts/parse-review.py` so bare `authentication` is no longer enough to classify `API_KEY_INVALID`.
- Kept timeout precedence intact and added regression coverage for timeout output that includes incidental auth wording.
- Made skip guidance error-specific for `RATE_LIMIT`, `SERVICE_UNAVAILABLE`, and generic `API_ERROR`.
- Added a walkthrough artifact and retro entry for the implementation lane.

## Alternatives Considered

- Do nothing: leaves developers chasing secrets for timeout/provider issues.
- Patch comment renderers only: rejected because the wrong structured error type would still propagate through artifacts and other consumers.
- Remove auth heuristics entirely: rejected because explicit unstructured auth failures without JSON would then lose useful classification.

## Acceptance Criteria

- [x] Given a reviewer that exits with code 124 and outputs `authentication`, the verdict is `timeout`, not `api_error`.
- [x] Given ambiguous auth-adjacent raw output without a timeout marker, the parser does not classify `authentication` alone as `API_KEY_INVALID`.
- [x] Given a `RATE_LIMIT` skip, downstream operator guidance no longer blames the API key.

## Manual QA

1. Run `python3 -m pytest tests/test_parse_review.py -q`
2. Run `python3 -m pytest tests/test_parse_review.py tests/test_render_findings.py tests/test_render_verdict_comment_helpers.py -q`
3. Run `make validate`
Expected:
- Targeted parser and rendering suites pass.
- `make validate` passes the full pytest phase (`1526 passed, 1 skipped`) and then fails on unrelated pre-existing `ruff` errors outside this PR.

## What Changed

Base branch flow:

```mermaid
flowchart TD
    A["Raw reviewer output"] --> B{"Timeout marker?"}
    B -->|No| C{"No JSON block?"}
    C -->|Yes| D{"Bare auth text present?"}
    D -->|Yes| E["SKIP: API_KEY_INVALID"]
    D -->|No| F["Generic parse failure"]
    B -->|Yes| G["SKIP: timeout"]
```

This PR flow:

```mermaid
flowchart TD
    A["Raw reviewer output"] --> B{"Timeout marker?"}
    B -->|Yes| G["SKIP: timeout"]
    B -->|No| C{"No JSON block?"}
    C -->|Yes| D{"Auth text + corroborating failure evidence?"}
    D -->|Yes| E["SKIP: API_KEY_INVALID"]
    D -->|No| F["Generic parse failure"]
```

Skip-message contract:

```mermaid
flowchart LR
    A["error_type"] --> B["generate_skip_verdict()"]
    B --> C["summary + title + suggestion"]
    C --> D["verdict artifacts"]
    C --> E["PR comment renderers"]
```

Why this shape is better:
- The parser now owns the truth about failure type instead of letting a loose substring leak into every downstream surface.
- Error-specific guidance is emitted once at the source and reused everywhere.

## Before / After

Before:
- Timeout or ambiguous parse-failure output could surface as `API_KEY_INVALID`.
- `RATE_LIMIT` skips told maintainers to check API keys.

After:
- Timeout markers reliably stay timeouts even when provider output mentions authentication.
- Ambiguous auth wording falls back to parse-failure unless corroborated.
- Rate-limit and service issues now produce operator guidance that matches the actual failure mode.

Walkthrough artifact:
- https://github.com/misty-step/cerberus/blob/codex/issue-282-timeout-api-error/artifacts/pr-282-walkthrough.md

## Test Coverage

- `tests/test_parse_review.py`
  - Added timeout-vs-auth regression coverage.
  - Added ambiguous-auth parse-failure coverage.
  - Added rate-limit guidance regression coverage.
- `python3 -m pytest tests/test_parse_review.py tests/test_render_findings.py tests/test_render_verdict_comment_helpers.py -q`

Gap:
- No new end-to-end GitHub Actions run was executed from this branch.

## Merge Confidence

Medium-high.

Strongest evidence:
- Focused RED -> GREEN tests for the exact parser failure mode.
- Full `tests/test_parse_review.py` pass.
- Broader rendering-related regression pass.
- Full repo pytest phase passed inside `make validate`.

Residual risk:
- Repo-wide lint remains red because of unrelated pre-existing files, so the full standard quality gate is not green yet.

## Summary by CodeRabbit

* **Bug Fixes**
  * Timeout detection now takes precedence over incidental authentication wording.
  * Tightened authentication heuristics to avoid false API-key classifications.
  * Improved rate-limit and service-unavailable guidance so SKIPs no longer point users to API keys.

* **Tests**
  * Added comprehensive tests for timeout precedence, auth-edge cases, rate-limit/quotas, and related fallback paths.

* **Documentation**
  * Added walkthrough, detailed PR artifact, and a retrospective entry describing the changes and verification steps.

### CODE REVIEW

File: scripts/parse-review.py [hunk 1]
```diff
@@ -442,7 +447,12 @@ def looks_like_api_error(text: str) -> tuple[bool, str, str]:
         (r"rate limit", "RATE_LIMIT", "Rate limit exceeded"),
         (r"quota exceeded", "API_CREDITS_DEPLETED", "API quota exceeded"),
         (r"billing", "API_CREDITS_DEPLETED", "Billing/quota error"),
-        (r"authentication", "API_KEY_INVALID", "Authentication error"),
+        (
+            r"authentication.{0,80}(?:401|403|failed|failure|invalid|incorrect|unauthorized)"
+            r"|(?:401|403|failed|failure|invalid|incorrect|unauthorized).{0,80}authentication",
+            "API_KEY_INVALID",
+            "Authentication error",
+        ),
```
[the remark this answers is not in the document]
Comment (phrazzld) [reply scripts/parse-review.py]: Classification: style
Severity: low
Decision: fix now. The old compound regex carried dead `401`/`403` branches.
Change: Commit `308cf0e` removes the compound auth tuple entirely and centralizes auth detection in `has_explicit_auth_signal()`, so there is a single auth rule instead of duplicated regex branches.
Verification: `/Users/phaedrus/.asdf/installs/python/3.12.8/bin/python3 -m pytest tests/test_parse_review.py tests/test_render_findings.py tests/test_render_verdict_comment_helpers.py -q`; `ruff check scripts/parse-review.py tests/test_parse_review.py`
